### DISCUSSION TITLE
Why this fragment makes this object `null` and the query not?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#2103] [Q&A] [answered]
I have a reproduction: https://github.com/frederikhors/iss-urql-fragment-null-obj

😃😃😃😃

### The problem

When you navigate to `/` it query the `Player:1` object.

When you navigate to `/team/1` it query the `Team:1` object.

If you goback to `/` the `Player` is `null` now!

If you change the [`PLAYER_FRAGMENT`](https://github.com/frederikhors/iss-urql-fragment-null-obj/blob/master/src/lib/graphql/queries/player.graphql) moving these lines:

```graphql
team {
    ...TEAM_FRAGMENT
}
```

to `query player()` it works (not `null` anymore).

### The steps

- `git clone https://github.com/frederikhors/iss-urql-fragment-null-obj`
- `npm install`
- go to https://codesandbox.io/s/apollo-server-urql-svelte-crud-h1pcl to start the server
- `npm run dev`

### The gif
![issue](https://user-images.githubusercontent.com/41120635/143720552-de19bca4-408c-4421-873d-3540fd4b5bc6.gif)

# Original question

If I use the below code it doesn't work, the second query (`team()`) makes the player `null` in cache.

```graphql
fragment PLAYER on Player {
  id
  createdAt
  team {
    ...TEAM
  }
}

fragment TEAM on Team {
  id
  createdAt
}

query player($id: ID!) {
  player(id: $id) {
    ...PLAYER
  }
}

query team($id: ID!) {
  team(id: $id) {
    ...TEAM
    player {
      ...PLAYER
    }
  }
}
```

The code below addresses the issue (this is the only change I do in my code base):

```graphql
fragment PLAYER on Player {
  id
  createdAt
  # removed these lines from here:
  # team {
  # 	...TEAM
  # }
}

fragment TEAM on Team {
  id
  createdAt
}

query player($id: ID!) {
  player(id: $id) {
    ...PLAYER
    # to here:
    team {
      ...TEAM
    }
  }
}

query team($id: ID!) {
  team(id: $id) {
    ...TEAM
    player {
      ...PLAYER
    }
  }
}
```

My question is: why it works in queries and not in fragments?

**Comment by frederikhors** [comment, marked as answer]
Sorry if I poke you guys, @JoviDeCroock @kitten (I'm in production with this mess), I can't even create a working codesandbox because it's really a mess to make them work there (I tried, really, you know me).

Have you ever had a fragment with an internal fragment that made an object `null` and instead using that same fragment in a query works (object not `null`)?

**Reply by JoviDeCroock** [reply]
I've managed to reproduce it in https://github.com/FormidableLabs/urql/pull/2107 will try to look at it today but can't promise anything.

**Reply by JoviDeCroock** [reply]
Eh.... Nevermind, this doesn't reproduce it I was being a dum-dum

**Reply by frederikhors** [reply]
I think I found a cause. I have this situation right now:

```graphql
fragment PLAYER on Player {
  id
  createdAt
  team {
    ...TEAM
  }
}

fragment TEAM on Team {
  id
  createdAt
}

query player($id: ID!) {
  player(id: $id) {
    ...PLAYER
  }
}

query team($id: ID!) {
  team(id: $id) {
    ...TEAM
    player {
    	...PLAYER
    }
  }
}
```

If I remove these lines:

```graphql
query team($id: ID!) {
  team(id: $id) {
    ...TEAM
    # player {
    # 	...PLAYER
    # }
  }
}
```

**It works!**

Can this help understand better?

**Reply by JoviDeCroock** [reply]
Because the reverse link isn't obvious. You need a resolver to say that team:x.player-->player:x because we only know the inverse player:x.team-->team:x

**Reply by frederikhors** [reply]
Oh. Wow.

Good news: it's not a bug.

Bad news: I did not understand what to do and what I did not read of the documentation to be so miserably wrong.

**Reply by frederikhors** [reply]
> You need a resolver to say that team:x.player-->player:x because we only know the inverse player:x.team-->team:x

Cannot this be automatic? I have a full schema, why I need to write these manually?

**Reply by JoviDeCroock** [reply]
Well it's hard for us to assume this as there could be fieldArguments and other things involved that make this a non-deterministic relationship.

**Reply by frederikhors** [reply]
Can you please show me an example of resolvers to write?

I did not understand this correctly.

If I use

```graphql
team {
  ...TEAM
}
```

in query it works. So the problem is the fragment?

**Reply by JoviDeCroock** [reply]
But then you don't have the 'team' right? Querying the player.team is what results in your odd behavior or am I seeing this wrong? Might help to show this in a bit more detail and show/tell what you see as these examples are deviating a lot now.

Query 1 data and fields
Query 2 data and fields

What are you seeing what is unexpected

**Reply by frederikhors** [reply]
The correct example is https://github.com/FormidableLabs/urql/discussions/2103#discussioncomment-1708881.

If I use that `...TEAM` and `...PLAYER` in queries instead of fragments everything works. No `null` anymore.

This is the reason I don't understand when you talk about deterministic links.

If this works with queries why don't with fragments?

**Reply by frederikhors** [reply]
@JoviDeCroock the final recap is here: https://github.com/FormidableLabs/urql/discussions/2103#discussion-3707904

**Reply by frederikhors** [reply]
@JoviDeCroock the question is updated with a REPRO. 😄

**Reply by JoviDeCroock** [reply]
So I tried looking at your repo which confused me quite a bit...

Your first query does the following, this is from the first / page:

```gql
query player($id: ID!) {
  player(id: $id) {
    ...PLAYER_FRAGMENT
  }
}

fragment PLAYER_FRAGMENT on Player {
  id
  description
  team {
    ...TEAM_FRAGMENT
  }
}

fragment TEAM_FRAGMENT on Team {
  id
  description
}
```

Which results in the output:

```json
{
   "data":{
      "player":{
         "id":"1",
         "description":"I'm a player",
         "team":{
            "id":"2",
            "description":"I'm a Team",
            "__typename":"Team"
         },
         "__typename":"Player"
      }
   }
}
```

We can see that we correctly have `Player:1.team` which points to `Team:2`, however when we switch to `/team/2`, the query that pops up is the following

```gql
query team($id: ID!) {
  team(id: $id) {
    ...TEAM_FRAGMENT
    player {
      ...PLAYER_FRAGMENT
    }
  }
}

fragment TEAM_FRAGMENT on Team {
  id
  description
}

fragment PLAYER_FRAGMENT on Player {
  id
  description
  team {
    ...TEAM_FRAGMENT
  }
}
```

And the response to that is

```json
{
   "data":{
      "team":{
         "id":"2",
         "description":"I'm a team",
         "player":{
            "id":"1",
            "description":"I'm a Player",
            "team":null,
            "__typename":"Player"
         },
         "__typename":"Team"
      }
   }
}
```

Do you see what's wrong here? You are effectively clearing `Player:1.team` to null because your API returns that...  This is all found out by leveraging the network-tab and calling `print()` on your AST documents. Hence why I asked for the responses rather than just looking at the queries.

When I alter your codesandbox to use GQL paradigms and handle this recursion we get to a correct result in your front-end https://codesandbox.io/s/apollo-server-urql-svelte-crud-forked-hbmvs?file=/schema.js:398-408 --> https://github.com/frederikhors/iss-urql-fragment-null-obj/pull/1

**Reply by frederikhors** [reply]
You're completely right. I had problems in my backend.
